File: app/src/main/java/com/aplicacion/mypet/activities/perfil/EditarPerfil.java
package com.aplicacion.mypet.activities.perfil;

import android.Manifest;
import android.app.Activity;
import android.app.AlertDialog;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.graphics.BitmapFactory;
import android.location.Address;
import android.location.Geocoder;
import android.net.Uri;
import android.os.Bundle;
import android.provider.MediaStore;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.ImageButton;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.core.content.FileProvider;

import com.aplicacion.mypet.R;
import com.aplicacion.mypet.fragments.BottomSheetFragmentPersonalizado;
import com.aplicacion.mypet.models.User;
import com.aplicacion.mypet.providers.AuthProvider;
import com.aplicacion.mypet.providers.ImageProvider;
import com.aplicacion.mypet.providers.UserProvider;
import com.aplicacion.mypet.utils.FileUtil;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.android.gms.tasks.Task;
import com.google.android.material.textfield.TextInputEditText;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.storage.UploadTask;
import com.squareup.picasso.Picasso;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

import de.hdodenhof.circleimageview.CircleImageView;
import dmax.dialog.SpotsDialog;

public class EditarPerfil extends AppCompatActivity  implements BottomSheetFragmentPersonalizado.BottomSheetListener {
    private final int GALLERY_REQUEST_CODE=1;
    private final int CAMARA_REQUEST_CODE=2;
    private final int REQUEST_PERMISSION_CAMERA=100;
    private final int REQUEST_PERMISSION_GALLERY=101;
    private AlertDialog dialog;
    private File imagen;
    private String nombre="";
    private boolean ocultarUbicacion;

    private ImageProvider imageProvider;
    private UserProvider userProvider;
    private String urlFotoPerfil ="";

    private Double latitude;
    private Double longitude;
    private final int REQUEST_CODE_UBICACION = 5;
    private Button botonUbicacion;

    private CircleImageView fotoPerfil;
    private TextInputEditText nombreUsuario;
    private ImageButton botonSubirFoto;

    private AuthProvider auth;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_editar_perfil);

        latitude = null;
        longitude = null;
        ocultarUbicacion = false;

        botonUbicacion = findViewById(R.id.boton_marcar_ubicacion);
        nombreUsuario = findViewById(R.id.campo_nombre_editar);
        fotoPerfil = findViewById(R.id.foto_perfil);
        botonSubirFoto = findViewById(R.id.boton_subir_foto);

        dialog = new SpotsDialog.Builder()
                .setContext(this)
                .setMessage(R.string.actualizando_perfil)
                .setCancelable(false).build();

        imageProvider = new ImageProvider();
        userProvider = new UserProvider();
        auth = new AuthProvider();

        getUser();
    }

    public void cerrarEditarPerfil(View view) {
        finish();
    }

    public void cambiarUbicacion(View view) {
        Intent ubicacion = new Intent(this,Ubicacion.class);
        if (latitude!=null && longitude!=null){
            ubicacion.putExtra("latitude", latitude);
            ubicacion.putExtra("longitude", longitude);
            ubicacion.putExtra("ocultarUbicacion",ocultarUbicacion);
        }
        startActivityForResult(ubicacion, REQUEST_CODE_UBICACION);
    }

    private void getUser(){
        userProvider.getUser(auth.getUid()).addOnSuccessListener(new OnSuccessListener<DocumentSnapshot>() {
            @Override
            public void onSuccess(DocumentSnapshot documentSnapshot) {
                if (documentSnapshot.exists()){

                    if (documentSnapshot.contains("username")) {
                        nombre = documentSnapshot.getString("username");
                        nombreUsuario.setText(nombre);
                    }
                    if (documentSnapshot.contains("ubicacion")){
                        ArrayList<Double> ubicacion = (ArrayList<Double>) documentSnapshot.get("ubicacion");
                        if (ubicacion!=null){
                            latitude = ubicacion.get(0);
                            longitude = ubicacion.get(1);
                            localizacion(latitude,longitude);
                        }
                    }

                    if (documentSnapshot.contains("ocultarUbicacion")) {
                        ocultarUbicacion = documentSnapshot.getBoolean("ocultarUbicacion");
                    }

                    if (documentSnapshot.contains("urlPerfil")){
                        urlFotoPerfil = documentSnapshot.getString("urlPerfil");
                        if (urlFotoPerfil!= null){
                            Picasso.get().load(urlFotoPerfil).into(fotoPerfil);
                        }
                    }

                }
            }
        });
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        if (requestCode==REQUEST_CODE_UBICACION){
            if (resultCode== Activity.RESULT_OK){
                Bundle extras = data.getExtras();
                latitude =  extras.getDouble("latitude");
                longitude = extras.getDouble("longitude");
                ocultarUbicacion = extras.getBoolean("ocultarUbicacion");

                localizacion(latitude,longitude);
            }
        } else if(requestCode==GALLERY_REQUEST_CODE && resultCode == RESULT_OK) {
            try {
                imagen = FileUtil.from(this,data.getData());

                fotoPerfil.setImageBitmap(BitmapFactory.decodeFile(imagen.getAbsolutePath()));
            }catch (Exception e){
                Log.d("ERROR: ", e.getMessage());
                Toast.makeText(this, "Error: " + e.getMessage(), Toast.LENGTH_SHORT).show();
            }
        } else if (requestCode == CAMARA_REQUEST_CODE && resultCode == RESULT_OK){
            fotoPerfil.setImageBitmap(BitmapFactory.decodeFile(imagen.getAbsolutePath()));
        }

        super.onActivityResult(requestCode, resultCode, data);
    }


    private void localizacion(double latitude, double longitude){
        Geocoder geocoder;
        List<Address> direccion = null;
        geocoder = new Geocoder(this, Locale.getDefault());

        try {
            direccion = geocoder.getFromLocation(latitude, longitude, 1); // 1 representa la cantidad de resultados a obtener
        } catch (IOException e) {
            e.printStackTrace();
        }

        String city = direccion.get(0).getLocality();
        String postalCode = direccion.get(0).getPostalCode();
        String country = direccion.get(0).getCountryName();
        String state = direccion.get(0).getLocality();

        if (postalCode!=null) {
            if (city!= null){
                botonUbicacion.setText(String.format("%s, %s", city, postalCode));
            } else {
                if (state!=null){
                    botonUbicacion.setText(String.format("%s, %s", state,country));
                } else {
                    botonUbicacion.setText(String.format("%s", country));
                }
            }
        } else {
            if (state!=null){
                botonUbicacion.setText(String.format("%s, %s", state,country));
            } else {
                botonUbicacion.setText(String.format("%s", country));
            }
        }
    }


    private void subirActualizacionPerfil() {
        if (imagen!=null) {
            imageProvider.save(EditarPerfil.this, imagen, 0).addOnSuccessListener(new OnSuccessListener<UploadTask.TaskSnapshot>() {
                @Override
                public void onSuccess(UploadTask.TaskSnapshot taskSnapshot) {
                    imageProvider.getStorage().addOnSuccessListener(new OnSuccessListener<Uri>() {
                        @Override
                        public void onSuccess(Uri uri) {
                            String url = uri.toString();
                            User usuario = new User();
                            usuario.setUsername(nombre);
                            usuario.setId(auth.getUid());
                            if (latitude != null && longitude != null) {
                                ArrayList<Double> ubicacion= new ArrayList<>();
                                ubicacion.add(latitude);
                                ubicacion.add(longitude);
                                usuario.setUbicacion(ubicacion);
                                usuario.setOcultarUbicacion(ocultarUbicacion);
                            }
                            usuario.setUrlPerfil(url);

                            userProvider.update(usuario).addOnCompleteListener(new OnCompleteListener<Void>() {
                                @Override
                                public void onComplete(@NonNull Task<Void> task) {
                                    if (task.isSuccessful()){
                                        dialog.dismiss();
                                        Toast.makeText(EditarPerfil.this, getString(R.string.perfil_actualizado), Toast.LENGTH_LONG).show();
                                        finish();
                                    } else {
                                        dialog.dismiss();
                                        Toast.makeText(EditarPerfil.this, getString(R.string.perfil_no_actualizado), Toast.LENGTH_LONG).show();
                                    }
                                }
                            });
                        }
                    });
                }
            });
        } else {
            User usuario = new User();
            usuario.setId(auth.getUid());
            if (!urlFotoPerfil.equals(""))
                usuario.setUrlPerfil(urlFotoPerfil);
            usuario.setUsername(nombre);
            if (latitude != null && longitude != null) {
                ArrayList<Double> ubicacion= new ArrayList<>();
                ubicacion.add(latitude);
                ubicacion.add(longitude);
                usuario.setUbicacion(ubicacion);
                usuario.setOcultarUbicacion(ocultarUbicacion);
            }
            userProvider.update(usuario).addOnCompleteListener(new OnCompleteListener<Void>() {
                @Override
                public void onComplete(@NonNull Task<Void> task) {
                    if (task.isSuccessful()){
                        dialog.dismiss();
                        Toast.makeText(EditarPerfil.this, getString(R.string.perfil_actualizado), Toast.LENGTH_LONG).show();
                        finish();
                    } else {
                        dialog.dismiss();
                        Toast.makeText(EditarPerfil.this, getString(R.string.perfil_no_actualizado), Toast.LENGTH_LONG).show();
                    }
                }
            });
        }
    }


    private void openGallery() {
        Intent galleryIntent = new Intent(Intent.ACTION_GET_CONTENT);
        galleryIntent.setType("image/*");
        startActivityForResult(galleryIntent,GALLERY_REQUEST_CODE);
    }

    private void openCamera() {
        Intent cameraIntent = new Intent(MediaStore.ACTION_IMAGE_CAPTURE);
        if(cameraIntent.resolveActivity(getPackageManager())!=null){

            imagen = null;
            try {
                imagen = FileUtil.fileCamera(this);
            } catch (IOException e) {
                e.printStackTrace();
            }

            if (imagen != null) {
                Uri fotoUri = FileProvider.getUriForFile(
                        this,
                        "com.aplicacion.mypet",
                        imagen);
                cameraIntent.putExtra(MediaStore.EXTRA_OUTPUT, fotoUri);
                startActivityForResult(cameraIntent,CAMARA_REQUEST_CODE);
            }
        }
    }

    @Override
    public void onButtonClicked(int numero) {
        if (numero == 0){
            if(ActivityCompat.checkSelfPermission(this, Manifest.permission.CAMERA) == PackageManager.PERMISSION_GRANTED){
                openCamera();
            } else {
                ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.CAMERA},REQUEST_PERMISSION_CAMERA);
            }
        } else if (numero == 1) {
            if(ActivityCompat.checkSelfPermission(this, Manifest.permission.READ_EXTERNAL_STORAGE) == PackageManager.PERMISSION_GRANTED){
                openGallery();
            } else {
                ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.READ_EXTERNAL_STORAGE},REQUEST_PERMISSION_GALLERY);
            }
        } else if (numero == 2) {
            fotoPerfil.setImageResource(R.drawable.ic_person_2);
            imagen=null;

        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        if(requestCode == REQUEST_PERMISSION_CAMERA) {
            if (permissions.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED){
                openCamera();
            } else {
                Toast.makeText(EditarPerfil.this, getString(R.string.permisos), Toast.LENGTH_LONG).show();
            }
        } else if(requestCode == REQUEST_PERMISSION_GALLERY) {
            if (permissions.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED){
                openGallery();
            } else {
                Toast.makeText(EditarPerfil.this, getString(R.string.permisos), Toast.LENGTH_LONG).show();
            }
        }
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
    }

    public void subirFoto(View view) {
        BottomSheetFragmentPersonalizado bottomSheet = new BottomSheetFragmentPersonalizado();
        bottomSheet.show(getSupportFragmentManager(),"TAG");
    }


    public void actualizarPerfil(View view) {
        nombre = nombreUsuario.getText().toString();
        if (!nombre.isEmpty()) {
            dialog.show();
            subirActualizacionPerfil();
        } else {
            Toast.makeText(this, getString(R.string.campos_vacios), Toast.LENGTH_LONG).show();
        }

    }


}

File: app/src/main/java/com/aplicacion/mypet/activities/perfil/EditarPerfil.kt
package com.aplicacion.mypet.activities.perfil

import android.Manifest
import android.app.AlertDialog
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.BitmapFactory
import android.location.Address
import android.location.Geocoder
import android.os.Bundle
import android.provider.MediaStore
import android.util.Log
import android.view.View
import android.widget.Button
import android.widget.ImageButton
import android.widget.Toast
import androidx.activity.result.contract.ActivityResultContracts.StartActivityForResult
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.core.content.FileProvider
import com.aplicacion.mypet.R
import com.aplicacion.mypet.fragments.BottomSheetFragmentPersonalizado
import com.aplicacion.mypet.models.User
import com.aplicacion.mypet.providers.AuthProvider
import com.aplicacion.mypet.providers.ImageProvider
import com.aplicacion.mypet.providers.UserProvider
import com.aplicacion.mypet.utils.FileUtil
import com.aplicacion.mypet.utils.ViewedMessageHelper
import com.google.android.material.textfield.TextInputEditText
import com.squareup.picasso.Picasso
import de.hdodenhof.circleimageview.CircleImageView
import dmax.dialog.SpotsDialog
import java.io.File
import java.io.IOException
import java.util.*

class EditarPerfil : AppCompatActivity(), BottomSheetFragmentPersonalizado.BottomSheetListener {
    private  val REQUEST_PERMISSION_CAMERA = 100
    private val REQUEST_PERMISSION_GALLERY = 101
    private lateinit var dialog: AlertDialog
    private var imagen: File? = null
    private lateinit var nombre: String
    private var ocultarUbicacion = false

    private lateinit var imageProvider: ImageProvider
    private lateinit var userProvider: UserProvider
    private var urlFotoPerfil : String? = ""

    private var latitude: Double? = null
    private var longitude: Double? = null
    private lateinit var botonUbicacion: Button

    private lateinit var fotoPerfil: CircleImageView
    private lateinit var nombreUsuario: TextInputEditText
    private lateinit var botonSubirFoto: ImageButton

    private lateinit var auth: AuthProvider

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_editar_perfil)


        ocultarUbicacion = false
        nombre = ""

        botonUbicacion = findViewById(R.id.boton_marcar_ubicacion)
        nombreUsuario = findViewById(R.id.campo_nombre_editar)
        fotoPerfil = findViewById(R.id.foto_perfil)
        botonSubirFoto = findViewById(R.id.boton_subir_foto)

        dialog = SpotsDialog.Builder()
                .setContext(this)
                .setMessage(R.string.actualizando_perfil)
                .setCancelable(false).build()

        imageProvider = ImageProvider()
        userProvider = UserProvider()
        auth = AuthProvider()

        getUser()
    }

    fun cerrarEditarPerfil(view: View?) {
        finish()
    }

    private fun getUser() {
        userProvider.getUser(auth.uid).addOnSuccessListener { documentSnapshot ->
            if (documentSnapshot.exists()) {
                if (documentSnapshot.contains("username")) {
                    nombre = documentSnapshot.getString("username")!!
                    nombreUsuario.setText(nombre)
                }
                if (documentSnapshot.contains("ubicacion")) {
                    val ubicacion = documentSnapshot["ubicacion"] as ArrayList<Double>?
                    if (ubicacion != null) {
                        latitude = ubicacion[0]
                        longitude = ubicacion[1]
                        localizacion(latitude!!, longitude!!)
                    }
                }
                if (documentSnapshot.contains("ocultarUbicacion")) {
                    if (documentSnapshot["ocultarUbicacion"] != null) ocultarUbicacion = documentSnapshot.getBoolean("ocultarUbicacion")!!
                }
                if (documentSnapshot.contains("urlPerfil")) {
                    urlFotoPerfil = documentSnapshot.getString("urlPerfil")
                    if (urlFotoPerfil != null)
                        Picasso.get().load(urlFotoPerfil).into(fotoPerfil)

                }
            }
        }
    }

    var resultUbicacion = registerForActivityResult(
            StartActivityForResult()
    ) { result ->
        if (result.resultCode == RESULT_OK) {
            val extras = result.data!!.extras
            latitude = extras!!.getDouble("latitude")
            longitude = extras.getDouble("longitude")
            ocultarUbicacion = extras.getBoolean("ocultarUbicacion")
            localizacion(latitude!!, longitude!!)
        }
    }

    var resultGaleria = registerForActivityResult(
            StartActivityForResult()
    ) { result ->
        if (result.resultCode == RESULT_OK) {
            try {
                imagen = FileUtil.from(this@EditarPerfil, result.data!!.data)
                fotoPerfil.setImageBitmap(BitmapFactory.decodeFile(imagen?.absolutePath))
            } catch (e: Exception) {
                Log.d("ERROR: ", e.message!!)
                Toast.makeText(this@EditarPerfil, "Error: " + e.message, Toast.LENGTH_SHORT).show()
            }
        }
    }

    var resultCamara = registerForActivityResult(
            StartActivityForResult()
    ) { result ->
        if (result.resultCode == RESULT_OK) {
            fotoPerfil.setImageBitmap(BitmapFactory.decodeFile(imagen?.absolutePath))
        }
    }

    fun cambiarUbicacion(view: View?) {
        val ubicacion = Intent(this, Ubicacion::class.java)
        if (latitude != null && longitude != null) {
            ubicacion.putExtra("latitude", latitude!!)
            ubicacion.putExtra("longitude", longitude!!)
            ubicacion.putExtra("ocultarUbicacion", ocultarUbicacion)
        }
        resultUbicacion.launch(ubicacion)
    }

    private fun localizacion(latitude: Double, longitude: Double) {
        val geocoder: Geocoder
        var direccion: List<Address>? = null
        geocoder = Geocoder(this, Locale.getDefault())
        try {
            direccion = geocoder.getFromLocation(latitude, longitude, 1) // 1 representa la cantidad de resultados a obtener
        } catch (e: IOException) {
            e.printStackTrace()
        }
        val city = direccion!![0].locality
        val postalCode = direccion[0].postalCode
        val country = direccion[0].countryName
        if (postalCode != null) {
            if (city != null) {
                botonUbicacion.text = String.format("%s, %s", city, postalCode)
            } else {
                botonUbicacion.text = String.format("%s", country)
            }
        } else {
            if (city != null) {
                botonUbicacion.text = String.format("%s, %s", city, country)
            } else {
                botonUbicacion.text = String.format("%s", country)
            }
        }
    }

    private fun subirActualizacionPerfil() {
        if (imagen != null) {
            if (urlFotoPerfil!= "")
                imageProvider.deleteByUrl(urlFotoPerfil)

            imageProvider.save(this@EditarPerfil, imagen, 0).addOnSuccessListener {
                imageProvider.storage.addOnSuccessListener { uri ->
                    val url = uri.toString()
                    val usuario = User()
                    usuario.username = nombre
                    usuario.id = auth.uid
                    if (latitude != null && longitude != null) {
                        val ubicacion = ArrayList<Double>()
                        ubicacion.add(latitude!!)
                        ubicacion.add(longitude!!)
                        usuario.ubicacion = ubicacion
                        usuario.ocultarUbicacion = ocultarUbicacion
                    }
                    usuario.urlPerfil = url
                    userProvider.update(usuario).addOnCompleteListener { task ->
                        if (task.isSuccessful) {
                            dialog.dismiss()
                            Toast.makeText(this@EditarPerfil, getString(R.string.perfil_actualizado), Toast.LENGTH_LONG).show()
                            finish()
                        } else {
                            dialog.dismiss()
                            Toast.makeText(this@EditarPerfil, getString(R.string.perfil_no_actualizado), Toast.LENGTH_LONG).show()
                        }
                    }
                }
            }
        } else {
            val usuario = User()
            usuario.id = auth.uid
            usuario.username = nombre
            if (latitude != null && longitude != null) {
                val ubicacion = ArrayList<Double>()
                ubicacion.add(latitude!!)
                ubicacion.add(longitude!!)
                usuario.ubicacion = ubicacion
                usuario.ocultarUbicacion = ocultarUbicacion
            }
            userProvider.update(usuario).addOnCompleteListener { task ->
                if (task.isSuccessful) {
                    dialog.dismiss()
                    Toast.makeText(this@EditarPerfil, getString(R.string.perfil_actualizado), Toast.LENGTH_LONG).show()
                    finish()
                } else {
                    dialog.dismiss()
                    Toast.makeText(this@EditarPerfil, getString(R.string.perfil_no_actualizado), Toast.LENGTH_LONG).show()
                }
            }
        }
    }

    private fun openGallery() {
        val galleryIntent = Intent(Intent.ACTION_GET_CONTENT)
        galleryIntent.type = "image/*"
        resultGaleria.launch(galleryIntent)
    }

    private fun openCamera() {
        val cameraIntent = Intent(MediaStore.ACTION_IMAGE_CAPTURE)
        if (cameraIntent.resolveActivity(packageManager) != null) {
            imagen = null
            try {
                imagen = FileUtil.fileCamera(this)
            } catch (e: IOException) {
                e.printStackTrace()
            }
            if (imagen != null) {
                val fotoUri = FileProvider.getUriForFile(
                        this,
                        "com.aplicacion.mypet",
                        imagen!!)
                cameraIntent.putExtra(MediaStore.EXTRA_OUTPUT, fotoUri)
                resultCamara.launch(cameraIntent)
            }
        }
    }

    override fun onButtonClicked(numero: Int) {
        if (numero == 0) {
            if (ActivityCompat.checkSelfPermission(this, Manifest.permission.CAMERA) == PackageManager.PERMISSION_GRANTED) {
                openCamera()
            } else {
                ActivityCompat.requestPermissions(this, arrayOf(Manifest.permission.CAMERA), REQUEST_PERMISSION_CAMERA)
            }
        } else if (numero == 1) {
            if (ActivityCompat.checkSelfPermission(this, Manifest.permission.READ_EXTERNAL_STORAGE) == PackageManager.PERMISSION_GRANTED) {
                openGallery()
            } else {
                ActivityCompat.requestPermissions(this, arrayOf(Manifest.permission.READ_EXTERNAL_STORAGE), REQUEST_PERMISSION_GALLERY)
            }
        } else if (numero == 2) {
            fotoPerfil.setImageResource(R.drawable.ic_person_2)
            imagen = null
        }
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String?>, grantResults: IntArray) {
        if (requestCode == REQUEST_PERMISSION_CAMERA) {
            if (permissions.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                openCamera()
            } else {
                Toast.makeText(this@EditarPerfil, getString(R.string.permisos), Toast.LENGTH_LONG).show()
            }
        } else if (requestCode == REQUEST_PERMISSION_GALLERY) {
            if (permissions.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                openGallery()
            } else {
                Toast.makeText(this@EditarPerfil, getString(R.string.permisos), Toast.LENGTH_LONG).show()
            }
        }
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
    }

    fun subirFoto(view: View?) {
        val bottomSheet = BottomSheetFragmentPersonalizado()
        bottomSheet.show(supportFragmentManager, "TAG")
    }

    fun actualizarPerfil(view: View?) {
        nombre = nombreUsuario.text.toString()
        if (!nombre.isEmpty()) {
            dialog.show()
            subirActualizacionPerfil()
        } else {
            Toast.makeText(this, getString(R.string.campos_vacios), Toast.LENGTH_LONG).show()
        }
    }

    override fun onStart() {
        super.onStart()
        if (auth.auth.currentUser != null) {
            ViewedMessageHelper.updateOnline(true, this)
        }
    }

    override fun onPause() {
        super.onPause()
        if (auth.auth.currentUser != null) {
            ViewedMessageHelper.updateOnline(false, this)
        }
    }
}